test/utils/messageSigner.ts

```diff
@@ -0,0 +1,140 @@
+import {Provider} from '@ethersproject/abstract-provider';
+import {Contract} from '@ethersproject/contracts';
+import {Web3Provider} from '@ethersproject/providers';
+import ethSigUtil, {TypedMessage} from '@metamask/eth-sig-util';
+import {BigNumber} from 'ethers';
+import {hexZeroPad} from 'ethers/lib/utils';
+
+/**
+ * Field in a User Defined Types
+ */
+export interface EIP712StructField {
+  name: string;
+  type: string;
+}
+
+/**
+ * User Defined Types are just an array of the fields they contain
+ */
+export type EIP712Struct = EIP712StructField[];
+/**
+ * Interface of the EIP712Domain structure
+ */
+export interface EIP712Domain {
+  name: string;
+  version: string;
+  chainId?: number;
+  verifyingContract: string;
+  salt?: string;
+}
+
+/**
+ * Interface of the complete payload required for signing
+ */
+export interface EIP712Payload {
+  types: PayloadTypes;
+  primaryType: string;
+  message: Record<string, string | number>;
+  domain: EIP712Domain;
+}
+
+export interface EIP712Signature {
+  hex: string;
+  v: number;
+  s: string;
+  r: string;
+}
+
+const EIP712Domain = [
+  {name: 'name', type: 'string'},
+  {name: 'version', type: 'string'},
+  {name: 'verifyingContract', type: 'address'},
+  {name: 'salt', type: 'bytes32'},
+];
+
+const ForwardRequest = [
+  {name: 'to', type: 'address'},
+  {name: 'from', type: 'address'},
+  {name: 'authorizer', type: 'address'},
+  {name: 'nftContract', type: 'address'},
+  {name: 'nonce', type: 'uint256'},
+  {name: 'nftChainId', type: 'uint256'},
+  {name: 'nftTokenId', type: 'uint256'},
+  {name: 'targetChainId', type: 'uint256'},
+  {name: 'data', type: 'bytes'},
+];
+
+interface PayloadTypes {
+  EIP712Domain: EIP712Struct;
+  ForwardRequest: EIP712Struct;
+}
+
+function getMetaTxTypeData(
+  verifyingContract: string,
+  _chainId: number,
+  message: Record<string, string | number>,
+  name: string
+): EIP712Payload {
+  return {
+    types: {
+      EIP712Domain,
+      ForwardRequest,
+    },
+    domain: {
+      name,
+      version: '0.0.1',
+      verifyingContract,
+      salt: hexZeroPad(BigNumber.from(_chainId).toHexString(), 32),
+    },
+    primaryType: 'ForwardRequest',
+    message,
+  };
+}
+
+async function signTypedData(signer: Web3Provider | string | Provider, from: string, data: EIP712Payload) {
+  // If signer is a private key, use it to sign
+  if (typeof signer === 'string') {
+    const privateKey = Buffer.from(signer.replace(/^0x/, ''), 'hex');
+    return ethSigUtil.signTypedData({
+      privateKey,
+      data: data as unknown as TypedMessage<any>,
+      version: ethSigUtil.SignTypedDataVersion.V3,
+    });
+  }
+
+  return await (signer as Web3Provider).send('eth_signTypedData_v4', [from, JSON.stringify(data)]);
+}
+
+async function attachNonce(forwarder: Contract, input: Record<string, any>): Promise<Record<string, any>> {
+  const nonce = await forwarder.getNonce(input.from).then((nonce: BigNumber) => nonce.toString());
```

Comment:
You're pulling this from the contract, which is great. But then you're just ignoring the one that was submitted. I don't think there's a problem with that, but then you can just drop it from the calldata generated by the preflight. Alternatively, you could pull this here and just use it as a check, and return an error if they don't match.

Reply:
trying to unpack this a bit!

i think perhaps this is just not a good test, where we are providing a nonce to the util function because we kinda just know what it is because the test environment?

I think really we just need to remove that from the test?

The forwarder includes nonce in calldata. Then the RPC fetches the current nonce, and refuses to sign the message if the provided nonce doesn't match the current nonce - https://github.com/0xEssential/playsession-rpc/pull/1/files#diff-1ba718c1eb8aa39cd20c2562d92523068c734d75f54655e97d652b992d9b4259R80

Perhaps you're right and this is all unnecessary? 

If the forwarder doesn't include nonce in calldata, then the ownership API will just sign with the current contract nonce. If we then verify the ownership signature using the state nonce, i think that would still keep us secure?

So yeah, we shouldn't provide nonce to the messageSigner in a dApp. And it feels like we can remove nonce from the error params and RPC check?